external-crates/move/move-compiler/src/typing/translate.rs

```diff
@@ -2265,3 +2267,161 @@ fn make_arg_types<S: std::fmt::Display, F: Fn() -> S>(
     }
     given
 }
+
+//**************************************************************************************************
+// Module-wide warnings
+//**************************************************************************************************
+
+/// Generates warnings for unused struct types and unused (private) functions.
+fn gen_unused_warnings(context: &mut Context, ident: ModuleIdent, mdef: &T::ModuleDefinition) {
+    let mut packed_types = BTreeSet::new();
+    let mut called_fns = BTreeSet::new();
+
+    for (_, _, fun) in &mdef.functions {
+        let sp!(_, body) = &fun.body;
+        if let T::FunctionBody_::Defined(seq) = body {
+            for item in seq {
+                let sp!(_, seq_item) = item;
+                seq_item_collect_used(&seq_item, &ident, &mut called_fns, &mut packed_types)
+            }
+        }
+    }
+
+    context
+        .env
+        .add_warning_filter_scope(mdef.warning_filter.clone());
+
+    for (loc, name, fun) in &mdef.functions {
+        context
+            .env
+            .add_warning_filter_scope(fun.warning_filter.clone());
+        if !called_fns.contains(name) {
+            // TODO: postponing handling of friend functions until we decide what to do with them
+            // vis-a-vis ideas around package-private
+            if let Visibility::Internal = fun.visibility {
+                if fun.entry.is_none() {
+                    let msg = "This function should be used or deleted.";
+                    context
+                        .env
+                        .add_diag(diag!(UnusedItem::Function, (loc, msg)))
+                }
+            }
+        }
+        context.env.pop_warning_filter_scope();
+    }
+
+    for (loc, name, struct_def) in &mdef.structs {
+        context
+            .env
+            .add_warning_filter_scope(struct_def.warning_filter.clone());
+        if !packed_types.contains(name) {
+            let msg = "This struct type should be used or deleted.";
```

Comment:
I don't think we should make this particular error as a part of the compiler, given that all types are currently public. (nor do I really think it should be a lint for the same reason). 

I'd probably instead be more focused on un-accessed fields? Since those are private to the module and a bit mroe impactful. 

And it kind of solves this problem. If you have an unused struct `Foo`, with fields, e.g.
```
struct Foo { f: u64 }
```
You will get an unused field `f: u64` warning. 
BUT! 
```
struct Foo {}
```
will be fine since that is just a public type that isn't used locally

Reply:
It does make a certain amount of sense if you call these "uninstantiated" rather than "unused" - what can you really do with a type that's never instantiated?